### DISCUSSION TITLE
xml structure from api for existing long urls with unique urls set to true

### DISCUSSION THREAD

**Discussion by adnaanqureshi** [YOURLS/YOURLS#3512] [Q&A] [answered]
hello
when unique urls is true as default, the xml returned by the api for a long url that already exists is different to the structure of a new long url. this means an app needs to factor in this variation. is it not possible for the api to simply return the short url that already exists, in the same xml structure as for a new link?

**Comment by ozh** [comment]
Can you illustrate with examples please ?

**Comment by logusivam** [comment, marked as answer]
You're correct, and this is a well-known inconsistency in the YOURLS API. When a URL already exists and `YOURLS_UNIQUE_URLS` is `true`, the API returns a different XML structure than when a new URL is created — and any API consumer has to handle both shapes, which is an unnecessary burden. Here's exactly what's happening and how to fix it today without waiting for a core change.

---

## The structural difference you're seeing

**New URL response:**
```xml
<yourls-api>
  <simple>http://sho.rt/abc123</simple>
  <message>http://example.com added to database</message>
  <errorCode>0</errorCode>
  <url>
    <keyword>abc123</keyword>
    <url>http://example.com</url>
    <title>Example</title>
    <date>2024-01-01 00:00:00</date>
    <ip>127.0.0.1</ip>
  </url>
  <shorturl>http://sho.rt/abc123</shorturl>
  <statusCode>200</statusCode>
  <status>success</status>
</yourls-api>
```

**Already-exists response:**
```xml
<yourls-api>
  <simple>http://sho.rt/abc123</simple>
  <message>http://example.com already exists in database</message>
  <errorCode>0</errorCode>
  <shorturl>http://sho.rt/abc123</shorturl>
  <statusCode>200</statusCode>
  <status>success</status>
</yourls-api>
```

The `<url>` block containing `<keyword>`, `<url>`, `<title>`, `<date>`, and `<ip>` is present in the new-URL response but **absent** in the already-exists response. The `<shorturl>` field is present in both, but the overall shape is different enough that naive XML parsing against a fixed schema fails.

---

## Workaround 1 — Fix this client-side (simplest, no server change needed)

Since `<shorturl>` is consistently present in **both** response shapes, the most straightforward fix is to always read from `<shorturl>` and ignore the `<url>` block for the purpose of getting the short URL — only accessing `<url>` fields when you specifically need keyword/title/date metadata:

```python
# Python example — works for both new and existing URL responses
import xml.etree.ElementTree as ET

root = ET.fromstring(xml_response)
short_url = root.findtext('shorturl')   # always present in both cases
status    = root.findtext('status')     # 'success' in both cases
message   = root.findtext('message')   # differs but both are informational

# Only access <url> block if it exists (new URL case only)
url_block = root.find('url')
keyword   = url_block.findtext('keyword') if url_block is not None else None
```

This pattern handles both XML shapes with no branching logic needed for the common case of "just give me the short URL."

---

## Workaround 2 — Normalize the response server-side via a plugin

If you want the API to return a **consistent XML structure** for both cases — identical to the new-URL response — you can hook into the YOURLS API response filter and inject the missing `<url>` block when the URL already exists.

Create `user/plugins/normalize-api-response/plugin.php`:

```php
<?php
/*
Plugin Name: Normalize API Response for Existing URLs
Plugin URI: https://yourls.org/
Description: Returns consistent XML/JSON structure for existing URLs (YOURLS_UNIQUE_URLS=true) matching the new-URL response shape
Version: 1.0
Author: You
*/

if ( !defined( 'YOURLS_ABSPATH' ) ) die();

yourls_add_filter( 'api_url_add', 'normalize_existing_url_response' );
function normalize_existing_url_response( $return ) {

    // Only act on the "already exists" case
    if (
        isset( $return['status'] ) && $return['status'] === 'success' &&
        isset( $return['message'] ) && strpos( $return['message'], 'already exists' ) !== false &&
        isset( $return['shorturl'] ) &&
        !isset( $return['url'] )   // the <url> block is missing — this is the inconsistent case
    ) {
        // Extract the keyword from the existing short URL
        $keyword = str_replace( YOURLS_SITE . '/', '', $return['shorturl'] );
        $keyword = yourls_sanitize_keyword( $keyword );

        // Fetch the full link data to populate the url block
        $link = yourls_get_keyword_infos( $keyword );

        if ( $link ) {
            $return['url'] = [
                'keyword' => $keyword,
                'url'     => $link['url'],
                'title'   => $link['title'],
                'date'    => $link['timestamp'],
                'ip'      => $link['ip'],
            ];
        }
    }

    return $return;
}
```

Install the usual way (`user/plugins/normalize-api-response/`) and activate from **Admin → Manage Plugins**.

After activation, the already-exists response will include the full `<url>` block, making both responses structurally identical:

```xml
<yourls-api>
  <simple>http://sho.rt/abc123</simple>
  <message>http://example.com already exists in database</message>
  <errorCode>0</errorCode>
  <url>
    <keyword>abc123</keyword>
    <url>http://example.com</url>
    <title>Example</title>
    <date>2024-01-01 00:00:00</date>
    <ip>127.0.0.1</ip>
  </url>
  <shorturl>http://sho.rt/abc123</shorturl>
  <statusCode>200</statusCode>
  <status>success</status>
</yourls-api>
```

This also normalizes the **JSON** response for the same reason — `format=json` callers benefit from the same fix automatically since both formats are generated from the same `$return` array.

---

## On whether this should be fixed in core

You're right that it should be. The inconsistency is not intentional design — it's an oversight in the original implementation of the duplicate-URL code path, which returns early with less data than the new-URL path builds up. A core fix would be straightforward: in `yourls-api.php` where the "already exists" branch returns, populate the `url` array the same way the new-URL branch does. This is worth opening as a bug report / PR against the YOURLS repo — the fix is small, the benefit to every API consumer is clear, and the plugin above is essentially the proof-of-concept implementation.

Hope this helps — if it answers your question, would you mind clicking **"Mark as answer"** so others searching for the same thing can find it quickly?
